blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,562 @@
+
+```typescript
+import { NodeSDK } from '@opentelemetry/sdk-node';
+import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-base';
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
+import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
+import { Resource } from '@opentelemetry/resources';
+import { SEMRESOURCENAME } from '@opentelemetry/semantic-conventions';
+import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+import { FetchInstrumentation } from '@opentelemetry/instrumentation-fetch';
+import { AlwaysOnSampler } from '@opentelemetry/sdk-trace-base';
+import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
+import { BatchLogRecordProcessor } from '@opentelemetry/sdk-logs';
+import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
+
+const exporter = new OTLPTraceExporter(config)
+const spanProcessor = new BatchSpanProcessor(exporter, opts)
+
+const logsExporter = new OTLPLogExporter(config)
+const logProcessor = new BatchLogRecordProcessor(logsExporter, opts)
+
+const metricsExporter = new OTLPMetricExporter(config)
+const metricsReader = new PeriodicExportingMetricReader({exporter: metricsExporter})
+
+// Configure the OTLP exporter to send data to your OpenTelemetry backend
+const config = { url: 'https://otel.highlight.io:4317' }
+const sdk = new NodeSDK({
+  autoDetectResources: true,
+  resourceDetectors: [processDetectorSync],
+  resource: new Resource({
+    [SEMRESOURCENAME.SERVICE_NAME]: 'nextjs-app',
+    'highlight.project_id': '<YOUR_PROJECT_ID>',
+  }),
+  spanProcessors: [spanProcessor],
+  logRecordProcessors: [logProcessor],
+  metricReader: metricReader,
+  traceExporter: exporter,
+  contextManager: new AsyncLocalStorageContextManager(),
+  sampler: new AlwaysOnSampler(),
+  instrumentations: [new HttpInstrumentation(), new FetchInstrumentation()],
+})
+
+sdk.start();
+```
+
+Now, let's use the OpenTelemetry SDK in our route to emit data:
+
+```typescript
+import { NextResponse } from "next/server";
+import api, { propagation } from "@opentelemetry/api";
+import { logs, SeverityNumber } from "@opentelemetry/api-logs";
+
+const tracerProvider = api.trace.getTracerProvider();
+const tracer = tracerProvider.getTracer("data");
+
+const loggerProvider = logs.getLoggerProvider();
+const logger = provider.getLogger("data");
+
+const meterProvider = api.metrics.getMeterProvider();
+const meter = meterProvider.getMeter("data");
+
+// This is an example implementation of a route that fetches data from a Python service
+export async function GET() {
+  const { email, name } = req.query;
+
+  console.log(`Fetching data for user: ${email} ${name}`);
+
+  // create a span for the data fetch
+  const data = await tracer.startActiveSpan(
```

Comment:
Don't we actually get this span automatically from `@opentelemetry/instrumentation-fetch`? Wondering if we should use a different example 🤔